external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private[streaming]
+  class DirectKafkaStreamingListener extends StreamingListener {
+
+    val offsetCommitEnabled = context.conf.getBoolean("spark.streaming.kafka.offsetCommit", false)
+
+    val zkClientOpt = if (offsetCommitEnabled) {
+      kafkaParams.get("zookeeper.connect").map { zkConnect =>
+        val zkSessionTimeoutMs = kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        val zkConnectionTimeoutMs =
+          kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        new ZkClient(zkConnect, zkSessionTimeoutMs, zkConnectionTimeoutMs, ZKStringSerializer)
+      }
+    } else {
+      None
+    }
+
+    val groupId = "directKafkaConsumer"
+
+    override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) = synchronized {
+      if (offsetCommitEnabled) {
+        for (offsets <- offsetMap.get(batchCompleted.batchInfo.batchTime); client <- zkClientOpt) {
```

Comment:
if offsetCommitEnabled, zkClientOpt should always be defined, yes?  Why two separate checks?

From a style point of view, people were complaining about monadic syntax earlier, I don't know that semicolon separated on a single line is an improvement.